Block Decorations

as-cii opened the pull request:
Implements and closes #9819.

![block-decorations](https://cloud.githubusercontent.com/assets/482957/11559474/6c857aea-99b9-11e5-921f-ba83bad27c09.gif)

The above GIF lasts `~ 50s` and showcases the integration of the block decorations API with the existing primitives within Atom (so, bear with it if it takes a bit to load :sweat_smile:). By default, we attach a decoration to a screen row: when the buffer changes or the underlying marker moves, we make sure to move the block decoration as well so that we can correctly calculate the other visual elements positions with respect to such decoration.

A block decoration can be created through the standard API to decorate markers:

``` js
editor.decorateMarker(
  editor.markScreenPosition([screenRow, 0]),
  {type: "block", item: item}
)
```

Decorations will be (re)measured if:
1. They are visible.
2. They are invisible but they were just added.
3. The editor width changes.
4. `TextEditorElement::invalidateBlockDecorationDimensions(blockDecoration)` is called.
5. They intersect with the buffer splicing region.

There could be some edge cases I haven't taken into account, but here's what I think it's missing:
- [x] Implement a non-linear (logarithmic) algorithm to handle screen-to-pixel positions conversion.
- [x] `offsetHeight` doesn't take margins into account.
- [x] Remove `addBlockDecorationForScreenRow`.
- [x] What happens when resizing the editor? Some block decorations above could need to be resized, thus changing all the items' position after them.
- [x] Correctly handle buffer vs. screen rows by invalidating the splicing region.
- [x] Handle custom gutters as well.
- [x] Inspect the `~ 3-4ms` slowdown when adding/removing many block decorations outside the shadow root.
  - This doesn't seem to be related to block decorations having a unique `id` or being teleported inside the shadow DOM because of `<content>`. Instead, it appears that some styles that cross the shadow boundary make style recalculation a bit slower when many block decorations get swapped in and out from screen. 
- [x] Add missing documentation (mark everything as experimental)
- [x] Allow to position block decorations before or after the marked position
- [x] Update documentation
- [ ] Test-drive! :car:
- [x] **Question:** should we implement a version that works without the shadow DOM? Maybe we can defer this to another PR, when the API stabilizes.

as-cii committed:
Change verticalScrollbar's height according to block decorations

as-cii committed:
:art: Create block decorations in `TextEditor`

This is just an experiment, although I like that we can hide some information
(irrelevant to the user) behind a clean API.

as-cii committed:
Ensure gutter scroll height takes block decorations into account

as-cii committed:
Ensure content scrollHeight takes block decorations into account

Seems like all these properties in the presenter's state share a common
contract, which we should probably extract. It could also mean there's an
underlying design problem, because we are testing two dimensions of the same
behavior in a single spec.

as-cii committed:
:art: Rename to ::blockDecorationsDimensionsById

as-cii committed:
:art: Rename to ::setBlockDecorationDimensions

as-cii committed:
Use ::rowForPosition and ::positionForRow instead of simple math

as-cii committed:
Use `::positionForRow` in `::updateTilesState`

The algorithm is still super sloooow :snail:, but we'll fix that in a later
commit.

as-cii committed:
:racehorse: Cache screen row height

I am trying to defer the usage of fancy algorithms as much as possible. The
current one is linear and should probably be changed, but it performs quite
decently for the time being. Maybe with some more caching we could even avoid to
implement a tree data structure?

as-cii committed:
Make sure tile positions are computed correctly

as-cii committed:
Remove caching by screen rows

Because we cannot use them as a cache key, because markers' position can change
at any time. Performance-wise this is slow with many markers, as we need to do a
lot of buffer-to-screen conversions.

as-cii committed:
Move position conversion in LinesYardstick

as-cii committed:
Use LinesYardstick consistently

as-cii committed:
Force cursor to be tall as ::lineHeight

as-cii committed:
Include blockDecorations inside each line's state

as-cii committed:
Provide blockDecorationsHeight for each line number

as-cii committed:
:art:

as-cii committed:
:art: Move block decoration related stuff into its own presenter

as-cii committed:
:art: Improve names a bit

as-cii committed:
:racehorse: Incremental updates for block decorations

as-cii committed:
:racehorse: Do as little pixel conversion as possible

We desperately need a tree-based data structure. :cry:

as-cii committed:
Use a boolean in each line state object

as-cii committed:
Include block decorations as a separate object on presenter's state

as-cii committed:
Start implementing BlockDecorationsComponent

as-cii committed:
Make sure cursors are updated with respect to block decorations

as-cii committed:
Implement block decorations in the components land

as-cii committed:
Implement a linear structure for block decoration coordinates

We still cannot handle `::rowForTopPixelPosition` when the passed top position
is in the middle of two rows and there's also a block decoration. We'll get
there eventually.

Also, the specs in this commit should serve as a good test suite for the future
logarithmic data structure.

as-cii committed:
Handle position between rows correctly

as-cii committed:
:fire: Remove old code

as-cii committed:
Use LinearLineTopIndex in BlockDecorationsPresenter

as-cii committed:
Use the new LineTopIndex in TextEditorPresenter

as-cii committed:
:art:

as-cii committed:
:bug: Coordinate conversion is hard

as-cii committed:
Write failing spec for measuring invisible elements

as-cii committed:
Use ::bottomPixelPositionForRow to scroll logically

as-cii committed:
Handle off-screen measurements properly

as-cii committed:
Use buffer marker events to avoid conversions

as-cii committed:
Render only visible and yet-to-be-measured block decorations

as-cii committed:
:racehorse: Use ids instead of classes

as-cii commented:
/cc: @nathansobo @maxbrunsfeld @atom/feedback

as-cii commented on spec/line-top-index-spec.js:
```diff
@@ -0,0 +1,189 @@
+/** @babel */
```
Maybe the specs here could benefit from some property (randomized) tests, as we do for other data structures such as the marker-index?

as-cii commented on src/block-decorations-presenter.js:
```diff
@@ -0,0 +1,150 @@
+const LineTopIndex = require('./linear-line-top-index')
+
+module.exports =
+class BlockDecorationsPresenter {
```
I opted to move block decorations handling out of `TextEditorPresenter` because it seemed like this code felt quite tight and cohesive. I am a bit unsure about the naming, though, because this doesn't seem to prepare any state for the view but rather to encapsulate information about block decorations + measurements and provide a cleaner API on top of them. Maybe this is just a `BlockDecorationsManager`? (although it feels a bit generic)

as-cii committed:
:fire:

as-cii committed:
More :fire:

as-cii committed:
Fix linting errors

philipgiuliani commented:
Are there still users that have not enabled the Shadow DOM? I already wondered why its not removed yet!

as-cii committed:
Ensure custom gutters work properly

as-cii committed:
:fire: :lipstick:

as-cii committed:
:green_heart: Fix false negative

as-cii commented on src/block-decorations-presenter.js:
```diff
@@ -0,0 +1,149 @@
+const {CompositeDisposable, Emitter} = require('event-kit')
+
+module.exports =
+class BlockDecorationsPresenter {
```
Putting a link to https://github.com/atom/atom/pull/9930#discussion_r46543078 which is not getting shown anymore because some commits were introduced after the comment.

as-cii committed:
Don't use babel when not needed

as-cii committed:
Do not remove invisible decorations on ::mouseWheelScreenRow

as-cii committed:
:art:

as-cii committed:
Make sure to add screen row to block decoration nodes

as-cii committed:
:green_heart: Fix specs

as-cii committed:
Make sure block decorations are always in the right spot

mostafaeweda commented on spec/text-editor-component-spec.js:
```diff
@@ -1651,6 +1651,106 @@ describe('TextEditorComponent', function () {
     })
   })
 
+  describe('block decorations rendering', function () {
```
since this will be transpiled by babel, why not preferring the `() => {}` over `function() {}`

as-cii replied:
Thanks for the suggestion, @mostafaeweda. I am not sure we have any convention to follow in this case, so I preferred consistency with other specs over syntax terseness. We could probably update all the other specs in another PR, although I have no strong preference over one syntax or the other, unless `this` needs to be captured.

mostafaeweda commented on src/linear-line-top-index.js:
```diff
@@ -0,0 +1,116 @@
+
+  topPixelPositionForRow (row) {
+    row = Math.min(row, this.maxRow)
+    let linesHeight = row * this.defaultLineHeight
```
revisit `let` --> should be `const` throughout the code.

mostafaeweda replied:
consider adding the eslint rules:
`prefer-const` & `no-const-assign` & `no-var`

mostafaeweda commented:
I can't express how awesome this is, especially that it aligns with Nuclide's Diff View (avoiding the screen lines to buffer lines hack to make hat happen).

as-cii committed:
:green_heart: Fix component and presenter specs

as-cii committed:
:green_heart: Fix LinesYardstick specs

as-cii committed:
Make sure screen row is set correctly on <content>

as-cii committed:
Merge branch 'master' into as-block-decorations

# Conflicts:
#	spec/text-editor-presenter-spec.coffee

as-cii committed:
:memo: TextEditor::addBlockDecorationForScreenRow

as-cii committed:
:memo: TextEditorElement::invalidateBlockDecorationDimensions

as-cii committed:
:white_check_mark: Test ::invalidateBlockDecorationDimensions

mikeball commented:
Nice work @as-cii, thank you so much for working on this!

as-cii committed:
Use an updated version of LineTopIndex

as-cii committed:
Start to move stuff back into TextEditorPresenter

as-cii committed:
Merge branch 'master' into as-block-decorations

# Conflicts:
#	spec/fake-lines-yardstick.coffee
#	spec/lines-yardstick-spec.coffee
#	spec/text-editor-presenter-spec.coffee
#	src/lines-yardstick.coffee
#	src/text-editor-component.coffee
#	src/text-editor-presenter.coffee

as-cii committed:
:green_heart: Adjust TextEditorComponent specs for block decorations

as-cii committed:
Start integrating tree-based LineTopIndex

as-cii committed:
Finish integrating LineTopIndex

as-cii committed:
:fire: Remove BlockDecorationsPresenter

as-cii committed:
:racehorse: Make updating block decorations incremental

as-cii committed:
Back to green specs :metal:

as-cii committed:
:fire: Remove maxRow parameter

as-cii committed:
Support only Shadow DOM enabled editors

as-cii committed:
Splice LineTopIndex when DisplayBuffer changes

We invalidate whole screen lines accordingly to `DisplayBuffer`, so that we can
catch if there was any screen-only transformation and move block decorations
accordingly.

as-cii committed:
:white_check_mark: Write specs for moving markers manually

as-cii committed:
Resize all block decorations when width changes

as-cii committed:
:fire: Remove TextEditor.prototype.addBlockDecorationForScreenRow

as-cii committed:
Take margin top and margin bottom into account

as-cii committed:
:arrow_up: line-top-index

as-cii committed:
:art: Rearrange code a bit

as-cii committed:
Invalidate spliced block decorations' dimensions

as-cii commented:
This seems to be fine and dandy after a test-drive. I will make sure to keep working on a build with these changes so that I can catch potential mistakes, but overall I feel like this is almost good to go.

@atom/feedback @mostafaeweda @mikeball @philipgiuliani: would love if someone could give this a :eyes: and poke around to ensure this works as expected.

@nathansobo @maxbrunsfeld: I think it's okay to :ship: this without support for non-shadow DOM editors. Adding such support would force us to introduce a different code path (which doesn't make use of `<content>` tags, etc.) and I feel like it's okay not add more complexity until this API stays experimental.

Feedback is super welcome! :tada:

mikeball commented:
Excellent I was able to get a very basic package(my first) working! A couple of things I noticed.

When the cursor is at the beginning of the line directly below the inserted block, and press enter to get a new line, the block is dragged down and new line inserted above the block. It would seem like the block should stay attached to the line above and the newline should be inserted below the block.

It would also seem logical that if you select text that contains a block with a standard text selection, and delete lines above and below, it should delete blocks contained within the selection?

Is it possible to make the text inside of the block item selectable? I tried various CSS rules but can’t seem to allow text contained in the block to be selectable with the mouse.

Thank you @as-cii !

as-cii committed:
Merge branch 'master' into as-block-decorations

as-cii commented:
> Is it possible to make the text inside of the block item selectable? I tried various CSS rules but can’t seem to allow text contained in the block to be selectable with the mouse.

I am not sure about that, but I feel like each package would need to handle this on its own. Block decorations are meant to be a very general purpose abstraction and, as such, I don't think that the core of Atom should provide anything that doesn't fit really common scenarios. Maybe selection is one of those scenarios, but I'd :heart: to see other examples before dealing with it.

> When the cursor is at the beginning of the line directly below the inserted block, and press enter to get a new line, the block is dragged down and new line inserted above the block. It would seem like the block should stay attached to the line above and the newline should be inserted below the block.

> It would also seem logical that if you select text that contains a block with a standard text selection, and delete lines above and below, it should delete blocks contained within the selection?

That's completely up to the marker you are decorating: the decoration will move with it and, therefore, it will follow any convention we have for markers. In this case, I think you'll want to listen for changes on that marker and move it somewhere else when it becomes invalid or your package's domain logic requires so.

mikeball commented:
>> When the cursor is at the beginning of the line directly below the inserted block, and press enter to get a new line, the block is dragged down and new line inserted above the block. It would seem like the block should stay attached to the line above and the newline should be inserted below the block.

> That's completely up to the marker you are decorating: the decoration will move with it and, therefore, it will follow any convention we have for markers. 

To clarify a bit what I was trying to say, when you insert a marker at say line 3, like this:
```coffeecript
  editor.markScreenPosition([3, 0]), {type: 'block', item: blockitem})
```
it's displayed above line 3 and moves along with line 3. I guess it would seem more natural for inline blocks to be displayed below line 3.

mikeball commented:
> > Is it possible to make the text inside of the block item selectable? 

> I am not sure about that, but I feel like each package would need to handle this on its own. Block decorations are meant to be a very general purpose abstraction and, as such, I don't think that the core of Atom should provide anything that doesn't fit really common scenarios. Maybe selection is one of those scenarios, but I'd :heart: to see other examples before dealing with it.

Also, completely agree, and it's great as is. I wasn't clear in asking, was trying to ask if it's possible for the package itself to make the contents selectable?

mostafaeweda commented:
I just tried replacing my diff view hack usage with this block decorations and it mostly worked fine.
However, when the editor has many block decorations and scrolling down fast, it errors:
```
Uncaught TypeError: Cannot read property 'blockHeight' of undefinedadjustNodeBlockHeight @ /Applications/Atom.app/Contents/Resources/app.asar/node_modules/line-top-index/dist/line-top-index.…:292resizeBlock @ /Applications/Atom.app/Contents/Resources/app.asar/node_modules/line-top-index/dist/line-top-index.…:97module.exports.TextEditorPresenter.setBlockDecorationDimensions @ /Applications/Atom.app/Contents/Resources/app.asar/src/text-editor-presenter.js:1645module.exports.BlockDecorationsComponent.measureBlockDecorations @ /Applications/Atom.app/Contents/Resources/app.asar/src/block-decorations-component.js:78module.exports.TextEditorComponent.readAfterUpdateSync @ /Applications/Atom.app/Contents/Resources/app.asar/src/text-editor-component.js:272__bind @ /Applications/Atom.app/Contents/Resources/app.asar/src/text-editor-component.js:3module.exports.ViewRegistry.performDocumentUpdate @ /Applications/Atom.app/Contents/Resources/app.asar/src/view-registry.js:219__bind @ /Applications/Atom.app/Contents/Resources/app.asar/src/view-registry.js:3
```
I'm suspecting this is a race condition in creating the map nodes: `blockEndNodesById` because this only happens the first time where I debugged and found the passed block item, but not the node entry in the map, then further fast scrolls are fine.

![screenshot 2015-12-28 11 41 37](https://cloud.githubusercontent.com/assets/264320/12024373/3f5ba10e-ad58-11e5-9ff6-30f1a413cddf.png)

as-cii committed:
:memo: Fix mistyped spec description

as-cii committed:
:fire: Remove redundant setup in specs

as-cii committed:
:bug: Ignore setting dimensions for destroyed decorations

as-cii committed:
:art:

as-cii committed:
Make spec more comprehensive

as-cii commented:
> I'm suspecting this is a race condition in creating the map nodes: blockEndNodesById because this only happens the first time where I debugged and found the passed block item, but not the node entry in the map, then further fast scrolls are fine.

@mostafaeweda: thanks so much for trying this out, I really appreciate it! :clap: The issue you have reported should have been fixed in c7a7f0c and I think it was caused by measuring a destroyed block  decoration which was rendered in a previous animation frame and hadn't been removed from the DOM yet.

> it's displayed above line 3 and moves along with line 3. I guess it would seem more natural for inline blocks to be displayed below line 3.

@mikeball: I am wondering if we should allow users to supply an option to specify whether the decoration should follow or precede the marked screen row. Specifically, I am thinking to cases where we want to support adding a block decoration before the first or after the last screen row. With the current approach, we cannot add a block decoration for the latter, which seems a bit unfortunate.

@nathansobo @maxbrunsfeld: thoughts? :thought_balloon:

nathansobo commented:
> I am wondering if we should allow users to supply an option to specify whether the decoration should follow or precede the marked screen row.

I think that makes sense.

as-cii committed:
Deal with rows instead of points in LineTopIndex

as-cii committed:
Add presenter specs for block decorations *after* a row

as-cii committed:
Conform to the new LineTopIndex interface

as-cii committed:
Split following and preceding block decorations

In TextEditorPresenter's state for lines.

as-cii committed:
Insert following block decorations into the DOM

as-cii committed:
Don't calculate height for following decorations at beginning of tiles

as-cii committed:
Slightly improve algorithm to calculate @endRow

as-cii committed:
:arrow_up: line-top-index

as-cii committed:
:memo: Include info for preceding/following block decorations

as-cii committed:
:green_heart: Give specs a final touch :sparkles:

as-cii commented:
@mikeball @mostafaeweda: we have now introduced the possibility of inserting block decorations after a specified row. I've updated the docs to describe this new feature: basically you now have the possibility of passing a `position` when decorating a marker: `'before'` if you want the decoration to be shown before the marked row, or `'after'` if the decoration should stick after the row. :rainbow: 

I feel like this is pretty stable and shouldn't really affect users/packages that do not interact with block decorations. Still, I'll let this bake for a while before :ship:ping it so it'd be really super awesome if you could try to break it during this period. :boom:

Thanks! :heart:

mostafaeweda commented:
Cool, will try it - but can't wait any longer to have it in an Atom release

as-cii committed:
Merge branch 'master' into as-block-decorations

as-cii committed:
Measure decorations only for visible editors
